### PR TITLE
Mailroom.py

### PR DESCRIPTION
**Pull request by Jaidjen**

### CODE REVIEW

File: students/Jaidjen/lesson03/Mailroom1.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
+import sys  # imports go at the top of the file
+
+
+Mail  =[('James Smith',[2000.25, 500, 27.50]),
```
Comment (natasha-aleksandrova) [inline students/Jaidjen/lesson03/Mailroom1.py]: this should be lower case snake_case > `mail`

I would also recommend naming this `donors` to more fit what this data holds

File: students/Jaidjen/lesson2/Series.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+nterms =10
+print("Fibonacci sequence:")
+for i in range(nterms):
+    print(fibonacci(i))
+
+
+def lucas(n):
+    if n == 0:
+        return 2;
+    if n == 1:
```
Comment (natasha-aleksandrova) [inline students/Jaidjen/lesson2/Series.py]: use `elif` here

File: students/Jaidjen/lesson03/Mailroom1.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    print("{:20s}{:20s}{:20s}{:20s}".format("DONOR ", "| TOTAL GIVEN ", "| NUMBER OF GIFTS ", "| AVERAGE GIFTS "))
+    print("-----------------------------------------------------------------------------")
+
+    for i in range(len(Mail)):
+        donate=[]
+        dnr_ave= sum(Mail[i][1])
+        ave_amt= int(len(Mail[i][1]))
+        tot_ave= dnr_ave/ave_amt
+        donate.append(dnr_ave)
+        sorted(donate)
```
Comment (natasha-aleksandrova) [inline students/Jaidjen/lesson03/Mailroom1.py]: this is expensive, you are sorting each time you're looping - you're also sorting and not using sorted data (you need to store what is returned by `sorted` in a variable.

File: students/Jaidjen/lesson03/Mailroom1.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+          "2 - Create a Report",
+          "3 - Quit",
+          ">>> "))
+
+
+def view_mail(Mail):
+    print("{:20s}{:20s}{:20s}{:20s}".format("DONOR ", "| TOTAL GIVEN ", "| NUMBER OF GIFTS ", "| AVERAGE GIFTS "))
+    print("-----------------------------------------------------------------------------")
+
+    for i in range(len(Mail)):
```
Comment (natasha-aleksandrova) [inline students/Jaidjen/lesson03/Mailroom1.py]: loop over the data structure directly, no need to use `range`

File: students/Jaidjen/lesson03/Mailroom1.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+         else:
+             dnr_amt = input("Please enter the donation amount: ")
+             Mail.append(new_mail)
+             Mail.append(dnr_amt)
+             print('Dear',new_mail,',')
+             print('Thank you', new_mail,'for donating','{:.2f}'.format(int(dnr_amt)))
+             print('We appreciate your generosity.')
+             print( )
+             print('Sincerely, Donation Team')
+             break
```
Comment (natasha-aleksandrova) [inline students/Jaidjen/lesson03/Mailroom1.py]: this function crashes when adding a donation:
```
Traceback (most recent call last):
  File "student.py", line 73, in <module>
    main()
  File "student.py", line 65, in main
    view_mail(Mail)
  File "student.py", line 24, in view_mail
    dnr_ave= sum(Mail[i][1])
TypeError: unsupported operand type(s) for +: 'int' and 'str'

```

File: students/Jaidjen/lesson03/Mailroom1.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        ave_amt= int(len(Mail[i][1]))
+        tot_ave= dnr_ave/ave_amt
+        donate.append(dnr_ave)
+        sorted(donate)
+        print("{:20s}{:^20.2f}{:^19d}{:^18.2f}".format(Mail[i][0], dnr_ave, len(Mail[i][1]), tot_ave))
+
+def add_mail():
+    print("If you want a list of donors please type 'list' ")
+    new_mail = input("Please enter the full name of the donor? ")
+    for i in range(len(Mail)):
```
Comment (natasha-aleksandrova) [inline students/Jaidjen/lesson03/Mailroom1.py]: this is not correct loop structure here - should be a while True - you are only using `i` when user selects "list"

File: students/Jaidjen/lesson03/Mailroom1.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+         if new_mail == 'list':
+             print("{:<20s}".format(Mail[i][0]))
+         elif new_mail == Mail[0]:
+            return Mail([0])
```
Comment (natasha-aleksandrova) [inline students/Jaidjen/lesson03/Mailroom1.py]: not sure what this code here is doing

File: students/Jaidjen/lesson03/Mailroom1.py (hunk 5 above)
Comment (natasha-aleksandrova) [reply students/Jaidjen/lesson03/Mailroom1.py]: this function should look more like this:
```
while True:
    response = input("")
    if response == "list":
        # list donors
    else:
        donation_amt = input("")
        for donor in donors:
            if donor[0] == response:
                # add donation
                break
        else:
            # new donor
```

File: students/Jaidjen/lesson04/Mailroom2.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
+import sys  # imports go at the top of the file
+
+
+Mail = {1:{'Name':'James Smith', 'Donation':[2000.25,500,27.50]},
+              2:{'Name': 'Dorothy Parker','Donation':[3175,475.45,92.30]},
+              3:{'Name':'Dohgyu Hwang','Donation':[10670,350.19,175]},
+              4:{'Name': 'Sally Yu', 'Donation':[100,925]},
+              5:{'Name':'Jenny Ramos','Donation':[25.14]}}
```
Comment (natasha-aleksandrova) [inline students/Jaidjen/lesson04/Mailroom2.py]: what is the benefit of using numbers for keys? the benefit of dictionaries is cheap look up by key with a natural key - your natural key here is the donor name.

File: students/Jaidjen/lesson04/Mailroom2.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+          "4 - Quit",
+          ">>> "))
+
+
+def view_mail(Mail):
+    print("{:20s}{:20s}{:20s}{:20s}".format("DONOR ", "| TOTAL GIVEN ", "| NUMBER OF GIFTS ", "| AVERAGE GIFTS "))
+    print("-----------------------------------------------------------------------------")
+    for mail in Mail:
+        sorted(Mail)
+        print("{:20s}{:^20.2f}{:^19d}{:^18.2f}".format(Mail[mail]['Name'],sum(Mail[mail]['Donation']), len(Mail[mail]['Donation']),(sum(Mail[mail]['Donation'])/(len(Mail[mail]['Donation'])))))
```
Comment (natasha-aleksandrova) [inline students/Jaidjen/lesson04/Mailroom2.py]: same issue with sorting

File: students/Jaidjen/lesson04/Mailroom2.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        dnr_amt = input("Please enter the donation amount: ")
+        print('Thank you', {new_mail}, 'for donating', {dnr_amt})
+    else:
+        dnr_amt = input("Please enter the donation amount: ")
+        Mail['Name']=new_mail
+        Mail['Donation']=dnr_amt
+        print('Dear',new_mail,',')
+        print('Thank you', new_mail,'for donating','{:.2f}'.format(int(dnr_amt)))
+        print('We appreciate your generosity.')
+        print('Sincerely, Donation Team')
```
Comment (natasha-aleksandrova) [inline students/Jaidjen/lesson04/Mailroom2.py]: still crashes

Comment (natasha-aleksandrova) [reply students/Jaidjen/lesson04/Mailroom2.py]: same issues with adding/updating donors. you should use the donor name as key and check if donor name is in your dictionary and add or update accordingly

File: students/Jaidjen/lesson04/Mailroom2.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        if response == "1":
+            add_mail()
+        elif response == "2":
+            view_mail(Mail)
+        elif response == "3":
+            thankyou_mail()
+        elif response == "4":
+            exit_program()
+        else:
+            print("Not a valid option!")
```
Comment (natasha-aleksandrova) [inline students/Jaidjen/lesson04/Mailroom2.py]: you should be using a switch dictionary here.

File: students/Jaidjen/lesson05/Mailroom3.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    new_mail = input("Please enter the full name of the donor? ")
+    if new_mail == 'list':
+        print("{:<20s}".format(Mail[Name]))
+    else:
+        try:
+            new_mail == Mail[1]
+            dnr_amt = input("Please enter the donation amount: ")
+        except ValueError:
+            print('Not acceptable')
+            raise ValueError
```
Comment (natasha-aleksandrova) [inline students/Jaidjen/lesson05/Mailroom3.py]: if you're catching the ValueError why raise it again? best to keep asking the user until valid value is entered.

File: students/Jaidjen/lesson05/Mailroom3.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        if response == "1":
+            add_mail()
+        elif response == "2":
+            view_mail(Mail)
+        elif response == "3":
+            thankyou_mail()
+        elif response == "4":
+            exit_program()
+        else:
+            print("Not a valid option!")
```
Comment (natasha-aleksandrova) [inline students/Jaidjen/lesson05/Mailroom3.py]: should be a switch dict here with exception handling.

**Comment by Jaidjen** [comment]
Hello,

Thank you for your input. I will work on fixing the issues this week and, if I need assistance, may I email you a request for OH? 

Thank you,
Diana

**Comment by Jaidjen** [comment]
Hello,

I can’t seem to connect zoom.

**Pull request merged**